### PR TITLE
Addfloats

### PR DESCRIPTION
**Pull request by btracey**
Sorry, still getting used to github. I requested to merge to the wrong branch. This should be better.

### CODE REVIEW

File: src/bench/floats/floats_test.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
+package floats
```
Comment (davecheney) [inline src/bench/floats/floats_test.go:1]: Please include a copyright statement and licence block

File: src/bench/floats/floats_test.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	benchmarkLogSumExp(b, s)
+}
+func BenchmarkLogSumExpHuge(b *testing.B) {
+	b.StopTimer()
+	s := RandomSlice(HUGE)
+	benchmarkLogSumExp(b, s)
+}
+
+func benchmarkDot(b *testing.B, s1 []float64, s2 []float64) {
+	b.StartTimer()
```
Comment (davecheney) [inline src/bench/floats/floats_test.go]: b.ResetTimer() then you don't need to remember who stopped or started. Drop all the other b.StopTimers() scattered around

File: src/bench/floats/floats_test.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
+package floats
+
+import (
+	"math"
+	"math/rand"
+	"testing"
+)
```
Comment (davecheney) [inline src/bench/floats/floats_test.go:7]: rand is not seeded.

Please seed it with a fixed seed

File: src/bench/floats/floats_test.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+// Performs a set of tests on simple mathematical operations with float slices
+
+// Add returns the element-wise sum of all the slices with the
+// results stored in the first slice.
+// For computational efficiency, it is assumed that all of
+// the variadic arguments have the same length. If this is
+// in doubt, EqLen can be used.
+func Add(dst []float64, slices ...[]float64) []float64 {
```
Comment (davecheney) [inline src/bench/floats/floats_test.go:20]: please gofmt this code

File: src/bench/floats/floats_test.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		// If it's infinity either way, the logsumexp will be infinity as well
+		// returning now avoids NaNs
+		return maxval
+	}
+	// Compute the sumexp part
+	for _, val := range s {
+		lse += math.Exp(val - maxval)
+	}
+	// Take the log and add back on the constant taken out
+	lse = math.Log(lse) + maxval
```
Comment (davecheney) [inline src/bench/floats/floats_test.go]: you don't need to assign to lse here, just return math.Log(lse) +maxval

File: src/bench/floats/floats_test.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+		if val > max {
+			max = val
+			ind = i
+		}
+	}
+	return max, ind
+}
+
+// Min returns the minimum value in the slice and the index of
+// the minimum value. If the input slice is empty, zero is returned
```
Comment (davecheney) [inline src/bench/floats/floats_test.go]: if len(s) == 0 then this function will panic, as Max does.

File: src/bench/floats/floats_test.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	SMALL       = 10
+	MEDIUM      = 1000
+	LARGE       = 100000
+	HUGE        = 10000000
+)
+
+func benchmarkMin(b *testing.B, s []float64) {
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		_, _ = Min(s)
```
Comment (davecheney) [inline src/bench/floats/floats_test.go]: just call Min(s)

File: src/bench/floats/floats_test.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	t := RandomSlice(i)
+	u := RandomSlice(i)
+	v := RandomSlice(i)
+	benchmarkAdd(b, s, t, u, v)
+}
+
+func benchmarkLogSumExp(b *testing.B, s []float64) {
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		_ = LogSumExp(s)
```
Comment (davecheney) [inline src/bench/floats/floats_test.go]: you don't need the assignment to _ here

File: src/bench/floats/floats_test.go (hunk 1 above)
Comment (btracey) [inline src/bench/floats/floats_test.go:1]: Should I copy the gonum license (BSD style) into the file, or should that go elsewhere?

File: src/bench/floats/floats_test.go (hunk 3 above)
Comment (btracey) [inline src/bench/floats/floats_test.go:7]: I can do that explicitly, but math/rand is by default a fixed seed

Comment (davecheney) [inline src/bench/floats/floats_test.go:7]: fair enough, it's documented so unlikely to change in the scope of Go 1.x, and 1 is as good a seed as any.

File: src/bench/floats/floats_test.go (hunk 1 above)
Comment (davecheney) [inline src/bench/floats/floats_test.go:1]: copyright at the top of the file, LICENCE file in the package

File: src/bench/floats/floats_test.go (hunk 4 above)
Comment (btracey) [inline src/bench/floats/floats_test.go:20]: Is there a flag to align the comments? Running gofmt didn't change the comments at all (nor the Add signature spacing)

Comment (davecheney) [inline src/bench/floats/floats_test.go:20]: I meant gofmt the whole file

Comment (davecheney) [inline src/bench/floats/floats_test.go:20]: please ignore this comment, for whatever reasons github shows the file misformatted, but the raw view shows that it is correct. Ignore my bleating to gofmt the code.

**Comment by btracey** [comment]
Pushed requested changes. PTAL

File: src/bench/floats/floats_test.go (hunk 1 above)
Comment (btracey) [inline src/bench/floats/floats_test.go:1]: Do the commits I made get automatically included with this merge request, or do I need to re-push it from github?

**Comment by davecheney** [comment]
Excellent. Thanks for bearing with me

**Comment by btracey** [comment]
No problem, thanks for bearing with me.

**Comment by btracey** [comment]
Sorry, there is one error. BenchmarkMinHuge did not have b.StopTimer() removed and so the benchmark will not stop. Should I make another change?

**Comment by davecheney** [comment]
Yup, new pr

**Comment by davecheney** [comment]
/home/dfc/devel/autobench/work/go.11/bin/go test  -test.run=XXX
-test.bench=. bench/floats > /home/dfc/devel/autobench/work/floats-11.txt
testing: warning: no tests to run
/home/dfc/devel/autobench/work/go.tip/bin/go test  -test.run=XXX
-test.bench=. bench/floats > /home/dfc/devel/autobench/work/floats-tip.txt
testing: warning: no tests to run
/home/dfc/devel/autobench/work/go.11/misc/benchcmp
/home/dfc/devel/autobench/work/floats-11.txt
/home/dfc/devel/autobench/work/floats-tip.txt
benchmark                  old ns/op    new ns/op    delta
BenchmarkMinSmall                 12           13   +0.78%
BenchmarkMinMed                  693          692   -0.14%
BenchmarkMinLarge              67208        66055   -1.72%
BenchmarkMinHuge             8936179      9025296   +1.00%
BenchmarkAddTwoSmall              21           25  +17.59%
BenchmarkAddFourSmall             43           48  +11.81%
BenchmarkAddTwoMed               982         1018   +3.67%
BenchmarkAddFourMed             2903         2929   +0.90%
BenchmarkAddTwoLarge          103809       101980   -1.76%
BenchmarkAddFourLarge         343695       342168   -0.44%
BenchmarkAddTwoHuge         16345354     16320656   -0.15%
BenchmarkAddFourHuge        49105734     49138748   +0.07%
BenchmarkLogSumExpSmall          265          264   -0.38%
BenchmarkLogSumExpMed          20086        20072   -0.07%
BenchmarkLogSumExpLarge      2015764      2012523   -0.16%
BenchmarkLogSumExpHuge     206482185    206366944   -0.06%
BenchmarkDotSmall                 15           16   +5.23%
BenchmarkDotMed                 1024         1032   +0.78%
BenchmarkDotLarge             106982       109836   +2.67%
BenchmarkDotHuge            13493495     13540756   +0.35%

**Comment by btracey** [comment]
Huh. I wonder why so many of those are positive (some significantly). Mine don't seem to agree on the most significant. (OSX 10.8.4).

brendan:~/Documents/mygo/src/github.com/davecheney/autobench/work/go.11/bin$ go test -test.run=XXX -test.bench=. github.com/davecheney/autobench/src/bench/floats > ~/Documents/mygo/src/github.com/davecheney/autobench/work/floats-11.txt
testing: warning: no tests to run
brendan:~/Documents/mygo/src/github.com/davecheney/autobench/work/go.11/bin$ go test -test.run=XXX -test.bench=. github.com/davecheney/autobench/src/bench/floats > ~/Documents/mygo/src/github.com/davecheney/autobench/work/floats-tip.txt
testing: warning: no tests to run
brendan:~/Documents/mygo/src/github.com/davecheney/autobench/work/go.11/misc$ ./benchcmp ~/Documents/mygo/src/github.com/davecheney/autobench/work/floats-11.txt ~/Documents/mygo/src/github.com/davecheney/autobench/work/floats-tip.txt
benchmark                  old ns/op    new ns/op    delta
BenchmarkMinSmall                 11           11   +0.00%
BenchmarkMinMed                  627          627   +0.00%
BenchmarkMinLarge              58201        58172   -0.05%
BenchmarkMinHuge             6767518      6731687   -0.53%
BenchmarkAddTwoSmall              19           19   +0.00%
BenchmarkAddFourSmall             38           38   +0.00%
BenchmarkAddTwoMed               875          874   -0.11%
BenchmarkAddFourMed             2607         2616   +0.35%
BenchmarkAddTwoLarge           87687        88812   +1.28%
BenchmarkAddFourLarge         264980       266091   +0.42%
BenchmarkAddTwoHuge         13052785     13648703   +4.57%
BenchmarkAddFourHuge        39087533     42606672   +9.00%
BenchmarkLogSumExpSmall          228          229   +0.44%
BenchmarkLogSumExpMed          17067        17131   +0.37%
BenchmarkLogSumExpLarge      1718625      1706672   -0.70%
BenchmarkLogSumExpHuge     172125533    171887143   -0.14%
BenchmarkDotSmall                 12           12   +0.00%
BenchmarkDotMed                  877          868   -1.03%
BenchmarkDotLarge              87429        86880   -0.63%
BenchmarkDotHuge            10669052     10819966   +1.41%

**Comment by davecheney** [comment]
I'd ignore anything below 3%, laptop cpu's reqiure too much thermal hand
holding to deliver constant results with a single sample. I'd concentrate
your energy on BenchmarkAddFourHuge

**Comment by davecheney** [comment]
BenchmarkAddTwoSmall              21           25  +17.59%
BenchmarkAddFourSmall             43           48  +11.81%

^ the sample size is too small, ignore these.

**Comment by btracey** [comment]
Wouldn't the sample size be much larger on the low-ns tests?

**Comment by davecheney** [comment]
We're only talking about a few ns, I don't trust the benchmark framework at
that resolution.

**Comment by btracey** [comment]
That's fair. I wouldn't necessarily either, I was just confused by your statement.

Thanks for merging the changes. If I have other code that I'm pretty sure is stable, I'll send more interesting benchmarks your way.

**Pull request merged**